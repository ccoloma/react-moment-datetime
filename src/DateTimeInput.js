import React from "react";
import Calendar from "./Calendar";
import { dateFormatToPattern, isDateValid } from "./util";
import { format, setHours, setMinutes, parse } from "date-fns/esm";
import { parseISO } from "date-fns";

let COUNTER = 0;
const DEFAULT_DATE_FORMAT = "yyyy-MM-dd";

// translate a date to the corresponding input values
function dateToValues(dateTimeStr, dateTimeFormat) {
  if (!dateTimeStr) {
    return {
      dateValue: "",
      timeValue: ""
    };
  }
  const date = parseISO(dateTimeStr);
  const [, dateFormat] = dateTimeFormat
    ? /^([^T]+)/.exec(dateTimeFormat)
    : [undefined, DEFAULT_DATE_FORMAT];
  return {
    dateValue: format(date, dateFormat),
    timeValue: format(date, "HH:mm")
  };
}

class DateTimeInput extends React.Component {
  constructor(props) {
    super(props);
    const { value, dateFormat } = props;
    if (value && typeof value !== "string") {
      throw new Error(
        `Property "value" must be of type String (received: ${value.constructor.name})`
      );
    }

    this.state = dateToValues(value, dateFormat);
    this.state.id = COUNTER++;

    this.onDateChange = e => {
      this.setState({
        dateValue: e.target.value
      });
    };

    this.onTimeChange = e => {
      this.setState({
        timeValue: e.target.value
      });
    };

    this.triggerChange = () => {
      const { isValid, dateFormat } = this.props;
      const { dateValue, timeValue } = this.state;
      let result = undefined;

      if (dateValue) {
        try {
          result = parse(dateValue, dateFormat, new Date());
          const timeParts = /([0-2]?[0-9]):([0-5]?[0-9])/.exec(timeValue);
          if (timeParts) {
            const [_, hour, minutes] = timeParts;
            result = setHours(result, +hour);
            result = setMinutes(result, +minutes);
          }
        } catch (e) {
          console.log(e.message);
          result = NaN;
        }
      }

      // notify of date change
      const onChange = this.props.onChange;
      isDateValid(result) &&
        (!isValid || isValid(result)) &&
        onChange &&
        onChange({
          name: this.props.name,
          value: result,
          strValue: !result
            ? undefined
            : !this.props.showTime
            ? format(result, "yyyy-MM-dd")
            : format(result, "yyyy-MM-dd'T'HH:mm")
        });
    };

    this.onInputFocus = () => {
      this.setState({
        focused: true,
        isOpen: true
      });
    };

    // moving focus to something else, close dialog
    this.onInputBlur = event => {
      const newElement = event.relatedTarget;
      if (newElement) {
        const isOpen = !!(
          newElement && newElement.closest(`#dt-${this.state.id}`)
        );
        this.setState({
          focused: false,
          isOpen: isOpen
        });
      }
      this.triggerChange();
    };

    // a date has been chosen in the calendar
    this.onCalendarChange = date => {
      const { dateFormat } = this.props;
      this.setState(
        {
          dateValue: format(date, dateFormat),
          isOpen: false
        },
        () => {
          setTimeout(() => {
            this.dateInputRef && this.dateInputRef.focus();
            this.setState({ isOpen: false });
          }, 0);
        }
      );
    };

    this.onInputClick = () => {
      this.setState({ isOpen: true });
    };

    this.onClose = () => {
      this.setState({ isOpen: false });
    };

    this.onKeyDown = e => {
      if (e.keyCode == 27) {
        // if esc, hide the calendar
        this.onClose();
      } else if (e.keyCode == 13) {
        // if enter is pressed when the calendar is shown, just cancel the event and hide the calendar
        // presing enter again will submit the form
        if (this.state.isOpen) {
          this.triggerChange();
          this.setState({ isOpen: false });
          e.preventDefault();
        }
      } else {
        // any other key, show the calendar
        this.state.isOpen || this.setState({ isOpen: true });
      }
    };
  }

  componentWillUnmount() {
    this.dateInputRef = undefined;
  }

  render() {
    const {
      i18n,
      dateFormat,
      isValid,
      showTime,
      required,
      name,
      disabled,
      placeholder,
      onChange,
      value,
      locale,
      ...otherProps
    } = this.props;
    const { isOpen, dateValue, timeValue, id } = this.state;

    return (
      <div className="dt-input-container" id={`dt-${id}`}>
        <div className="dt-inputs">
          <div className="dt-input-date-wrapper">
            <input
              type="text"
              className="dt-input dt-input-date"
              value={dateValue}
              required={required}
              placeholder={placeholder || dateFormat.toLowerCase()}
              pattern={dateFormatToPattern(dateFormat)}
              name={name}
              disabled={disabled}
              autoComplete="off"
              ref={e => (this.dateInputRef = e)}
              onKeyDown={!disabled ? this.onKeyDown : undefined}
              onClick={!disabled ? this.onInputClick : undefined}
              onFocus={!disabled ? this.onInputFocus : undefined}
              onBlur={!disabled ? this.onInputBlur : undefined}
              onChange={!disabled ? this.onDateChange : undefined}
              {...otherProps}
            />
          </div>

          {showTime && (
            <input
              type="text"
              className="dt-input dt-input-time"
              pattern="([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]"
              value={timeValue}
              required={required}
              disabled={disabled}
              placeholder="hh:mm"
              onBlur={!disabled ? this.onInputBlur : undefined}
              onChange={!disabled ? this.onTimeChange : undefined}
            />
          )}
        </div>

        {isOpen && (
          <div aria-hidden="true">
            <div className="dt-dialog-backdrop" onClick={this.onClose} />
            <dialog className="dt-dialog" open>
              <Calendar
                dateValue={dateValue}
                dateFormat={dateFormat}
                isValid={isValid}
                onChange={this.onCalendarChange}
                locale={locale}
              />
              <div className="dt-actions">
                <button
                  type="button"
                  className="dt-btn dt-close"
                  tabIndex="-1"
                  onClick={this.onClose}
                >
                  {i18n.Close}
                </button>
              </div>
            </dialog>
          </div>
        )}
      </div>
    );
  }
}

DateTimeInput.defaultProps = {
  // Date format
  dateFormat: DEFAULT_DATE_FORMAT,

  // the i18n entries to use
  i18n: {
    Close: "Close"
  },

  // true to include a time component
  showTime: true,

  // the current value, as date
  // value: undefined,

  // a function that given a date returns if it's valid
  isValid: date => true,

  // triggered when there is a change. Receives a Date instance
  onChange: undefined
};

export default DateTimeInput;
