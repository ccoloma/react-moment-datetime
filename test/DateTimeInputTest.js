import React from "react";
import renderer from "react-test-renderer";
import DateTimeInput from "../src/DateTimeInput";
import { parseISO, setDate } from "date-fns/esm";

describe("DateTimeInput", function() {
  const date = parseISO("2014-10-25T10:20Z");

  it("should render with empty date", function() {
    renderer.create(<DateTimeInput />);
  });

  it("should render with non-empty date", function() {
    const tree = renderer.create(<DateTimeInput value={date} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  function getDateInput(tree) {
    return tree.root.findByProps({
      className: "dt-input dt-input-date"
    });
  }

  function getTimeInput(tree) {
    return tree.root.findByProps({
      className: "dt-input dt-input-time"
    });
  }

  it("should respond to text introduced by the user", function() {
    const onChange = jest.fn();
    const tree = renderer.create(
      <DateTimeInput name="foobar" value={date} onChange={onChange} />
    );

    // edit date
    const dateInput = getDateInput(tree);
    const timeInput = getTimeInput(tree);

    dateInput.props.onFocus();
    dateInput.props.onChange({ target: { value: "2014-10-05" } });
    expect(onChange).toHaveBeenCalledTimes(0);
    dateInput.props.onBlur({});
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith({
      name: "foobar",
      strValue: "2014-10-05T12:20",
      value: setDate(date, 5)
    });

    // edit time
    timeInput.props.onChange({ target: { value: "17:10" } });
    dateInput.props.onBlur({});
    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenLastCalledWith({
      name: "foobar",
      strValue: "2014-10-05T17:10",
      value: parseISO("2014-10-05T17:10")
    });

    // empty date
    dateInput.props.onChange({ target: { value: "" } });
    dateInput.props.onBlur({});
    expect(onChange).toHaveBeenCalledTimes(3);
    expect(onChange).toHaveBeenLastCalledWith({
      name: "foobar",
      strValue: undefined,
      value: undefined
    });
  });

  it("should render a calendar", function() {
    const tree = renderer.create(<DateTimeInput value={date} />);
    const dateInput = getDateInput(tree);
    dateInput.props.onFocus();
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("handle invalid date", function() {
    const onChange = jest.fn();
    const tree = renderer.create(<DateTimeInput onChange={onChange} />);
    const dateInput = getDateInput(tree);
    dateInput.props.onFocus();
    dateInput.props.onChange({ target: { value: "xxx" } });
    dateInput.props.onBlur({});
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  it("should render disabled inputs", function() {
    const tree = renderer.create(
      <DateTimeInput value={date} disabled={true} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
