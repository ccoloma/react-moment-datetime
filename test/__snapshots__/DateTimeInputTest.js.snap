// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimeInput should render a calendar 1`] = `
<div
  className="dt-input-container"
  id="dt-3"
>
  <div
    className="dt-inputs"
  >
    <div
      className="dt-input-date-wrapper"
    >
      <input
        autoComplete="off"
        className="dt-input dt-input-date"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        pattern="[0-9]{4}-[01]?[0-9]-[0-3]?[0-9]"
        placeholder="yyyy-mm-dd"
        type="text"
        value="2014-10-25"
      />
    </div>
    <input
      className="dt-input dt-input-time"
      onBlur={[Function]}
      onChange={[Function]}
      pattern="([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]"
      placeholder="hh:mm"
      type="text"
      value="10:20"
    />
  </div>
  <div
    aria-hidden="true"
  >
    <div
      className="dt-dialog-backdrop"
      onClick={[Function]}
    />
    <dialog
      className="dt-dialog"
      open={true}
    >
      <div
        aria-hidden="true"
        className="dt-calendar"
      >
        <div
          className="dt-toolbar"
        >
          <button
            className="dt-button dt-btn-prev-month"
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          />
          <span
            className="dt-current-date"
          >
            October 2014
          </span>
          <button
            className="dt-button dt-btn-next-month"
            onClick={[Function]}
            tabIndex="-1"
            type="button"
          />
        </div>
        <table
          className="dt-calendar-table"
        >
          <thead>
            <tr>
              <th
                className="dt-th"
              >
                Sun
              </th>
              <th
                className="dt-th"
              >
                Mon
              </th>
              <th
                className="dt-th"
              >
                Tue
              </th>
              <th
                className="dt-th"
              >
                Wed
              </th>
              <th
                className="dt-th"
              >
                Thu
              </th>
              <th
                className="dt-th"
              >
                Fri
              </th>
              <th
                className="dt-th"
              >
                Sat
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a dt-other-month"
                  data-date="2014-09-28T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  28
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a dt-other-month"
                  data-date="2014-09-29T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  29
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a dt-other-month"
                  data-date="2014-09-30T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  30
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-01T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  1
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-02T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  2
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-03T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  3
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-04T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  4
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-05T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  5
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-06T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  6
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-07T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  7
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-08T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  8
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-09T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  9
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-10T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  10
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-11T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  11
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-12T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  12
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-13T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  13
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-14T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  14
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-15T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  15
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-16T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  16
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-17T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  17
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-18T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  18
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-19T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  19
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-20T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  20
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-21T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  21
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-22T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  22
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-23T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  23
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-24T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  24
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a dt-current-day"
                  data-date="2014-10-25T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  25
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-26T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  26
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-27T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  27
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-28T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  28
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-29T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  29
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-30T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  30
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a"
                  data-date="2014-10-31T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  31
                </a>
              </td>
              <td
                className="dt-day"
              >
                <a
                  className="dt-day-a dt-other-month"
                  data-date="2014-11-01T00:00:00.000Z"
                  onMouseDown={[Function]}
                  tabIndex="-1"
                >
                  1
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        className="dt-actions"
      >
        <button
          className="dt-btn dt-close"
          onClick={[Function]}
          tabIndex="-1"
          type="button"
        >
          Close
        </button>
      </div>
    </dialog>
  </div>
</div>
`;

exports[`DateTimeInput should render disabled inputs 1`] = `
<div
  className="dt-input-container"
  id="dt-5"
>
  <div
    className="dt-inputs"
  >
    <div
      className="dt-input-date-wrapper"
    >
      <input
        autoComplete="off"
        className="dt-input dt-input-date"
        disabled={true}
        pattern="[0-9]{4}-[01]?[0-9]-[0-3]?[0-9]"
        placeholder="yyyy-mm-dd"
        type="text"
        value="2014-10-25"
      />
    </div>
    <input
      className="dt-input dt-input-time"
      disabled={true}
      pattern="([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]"
      placeholder="hh:mm"
      type="text"
      value="10:20"
    />
  </div>
</div>
`;

exports[`DateTimeInput should render with non-empty date 1`] = `
<div
  className="dt-input-container"
  id="dt-1"
>
  <div
    className="dt-inputs"
  >
    <div
      className="dt-input-date-wrapper"
    >
      <input
        autoComplete="off"
        className="dt-input dt-input-date"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        pattern="[0-9]{4}-[01]?[0-9]-[0-3]?[0-9]"
        placeholder="yyyy-mm-dd"
        type="text"
        value="2014-10-25"
      />
    </div>
    <input
      className="dt-input dt-input-time"
      onBlur={[Function]}
      onChange={[Function]}
      pattern="([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]"
      placeholder="hh:mm"
      type="text"
      value="10:20"
    />
  </div>
</div>
`;
